### PR TITLE
Convert consumer to use TRY-EXCEPT for work item release

### PR DESCRIPTION
**Pull request by kylie-bee**
Update the consumer bot to use `TRY-EXCEPT` instead of `IF-ELSE` to detect failure and release work item appropriately. Also update the simulated error to be either `BUSINESS` or `APPLICATION` type to highlight how to catch different types of errors in one `TRY` block.

### CODE REVIEW

File: producer.robot [hunk 1]
```diff
@@ -11,7 +11,7 @@ Produce items
     ...    Get source Excel file from work item.
     ...    Read rows from Excel.
     ...    Creates output work items per row.
-    ${path}=    Get Work Item File    orders.xlsx    %{ROBOT_ARTIFACTS}${/}orders.xlsx
+    ${path}=    Get Work Item File    orders.xlsx    %{ROBOT_ARTIFACT=temp}${/}orders.xlsx
```
Comment (cmin764) [inline producer.robot:14]: I want to learn more: What's with the `ROBOT_ARTIFACT` env var here, as is the first time I'm seeing this? :D

Is it something injected by CR which points to a place from where the artifact files will be uploaded after the Step run finishes? If yes, then wouldn't be `${OUTPUT_DIR}` a better option here? (as an opportunistic refactoring)

**Review by cmin764 — CHANGES_REQUESTED** [review]
GJ on refactoring it into using TRY-EXCEPTs, the pattern is much more cleaner and robust now.

Although left a few concerns I have with the following:
- exception type on timeout
- env JSONs removal
- output path for the retrieved work item file

File: producer.robot (hunk 1 above)
Comment (blakewaud) [reply producer.robot:14]: I honestly had no idea where this env variable came from, but figured to avoid problems, I'd give it a default. The `env` files didn't have it before I deleted them and it's not a built in or automatic variable.

File: devdata/env-for-consumer.json [hunk 2]
```diff
@@ -1,5 +0,0 @@
-{
```
[the remark this answers is not in the document]
Comment (blakewaud) [reply devdata/env-for-consumer.json:1]: As far as I know, only `env.json` is automatically used by our tooling. I tried using `rcc` only to create a producer-consumer bot from the template and then run it, but it fails because the bot is expecting environment variable `RC_WORKSPACE_ID`, which tells me that it's not actually using those files and these are hold overs from how we used to simulate the cloud locally within VSCode before our extensions did it.

File: devdata/env.json [hunk 3]
```diff
@@ -1,5 +0,0 @@
-{
```
[the remark this answers is not in the document]
Comment (blakewaud) [reply devdata/env.json:1]: Agreed

File: consumer.robot [hunk 4]
```diff
@@ -35,33 +36,54 @@ Login
...
         ...    exception_type=BUSINESS
+        ...    code=INVALID_DATA
         ...    message=${error_message}
+    EXCEPT    *timed out*    type=GLOB    AS    ${err}
+        ${error_message}=    Set Variable
+        ...    Application error encountered: ${err}
+        Log    ${error_message}    level=ERROR
+        Release Input Work Item
+        ...    state=FAILED
+        ...    exception_type=BUSINESS
```
[the remark this answers is not in the document]
Comment (blakewaud) [reply consumer.robot:83]: Yup! Missed that in my copy pasta!

File: producer.robot (hunk 1 above)
Comment (blakewaud) [reply producer.robot:14]: Okay, I booted up the interactive console and was able to confirm that `%{ROBOT_ARTIFACTS}` is a built in environment variable (from our tooling maybe?) that points to the `${OUTPUT DIR}`. So this change of mine is unnecessary.

Comment (blakewaud) [reply producer.robot]: Agreed, and pushed

File: devdata/env-for-consumer.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (cmin764) [reply devdata/env-for-consumer.json:1]: Indeed, but you can run the robot with `rcc run -e path/to/any-env-name.json -t <task-name>` so we still need _devdata/env-for-consumer.json_ and _devdata/env-for-producer.json_ available so we can run the robot outside of VSCode without complaining about any `RC_WORKSPACE_ID`.

Can you try testing that on your side to confirm it is working? (then keep these two files)

----

The _devdata/env.json_ indeed is automatically detected by VSCode, but since the extension can now detect the presence of Work Items JSON files from _devdata_, we can safely remove **ONLY** the _devdata/env.json_ file. (and keep the rest)

**Review by cmin764 — APPROVED** [review]
Thx for the changes! Let's get one more with those env JSONs if they make sense to you. (then you can merge)

File: devdata/env-for-consumer.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (blakewaud) [reply devdata/env-for-consumer.json:1]: Okay yup, forgot about that cli flag, I will bring those two files back, thanks.

File: devdata/env.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (mikahanninen) [reply devdata/env.json:1]: love the amicable agreement here, but I think is **needed** because this one is used by default in VSCode when running with Robocorp Code extension unless I am totally mistaken

File: consumer.robot [hunk 5]
```diff
@@ -8,18 +8,19 @@ Library     RPA.Tables
...
 Consume items
     [Documentation]    Login and then cycle through work items.
-    ${Login OK}=    Run Keyword And Return Status
-    ...    Login
-    IF    ${Login OK}
-        For Each Input Work Item    Handle item
-    ELSE
+    TRY
+        Login
+    EXCEPT    Login failed
```
Comment (mikahanninen) [inline consumer.robot:13]: I think this should be bare EXCEPT without any message. We would want to catch all error situations with EXCEPT. I know that in our example we can rely on the error message, but this template should provide best practice and in complex keywords it is not wise to rely on an exact error to happen.

File: consumer.robot [hunk 6]
```diff
@@ -35,33 +36,54 @@ Login
...
 Handle item
     [Documentation]    Error handling around one work item.
-    ${payload}=    Get Work Item Payload
-    ${Item Handled}=    Run Keyword And Return Status
-    ...    Action for item    ${payload}
-    IF    ${Item Handled}
-        Release Input Work Item    DONE
-    ELSE
+    ${payload}=    Get Work Item Variables
+    TRY
```
Comment (mikahanninen) [inline consumer.robot:63]: This TRY-EXCEPT block is good for detecting specific errors but if any other error than expected 2 cased happens then those are not caught and this block will fail to handle that.

File: producer.robot [hunk 7]
```diff
@@ -11,7 +11,7 @@ Produce items
     ...    Get source Excel file from work item.
     ...    Read rows from Excel.
     ...    Creates output work items per row.
-    ${path}=    Get Work Item File    orders.xlsx    %{ROBOT_ARTIFACTS}${/}orders.xlsx
+    ${path}=    Get Work Item File    orders.xlsx    ${OUTPUT_DIR}${/}orders.xlsx
```
Comment (mikahanninen) [inline producer.robot:14]: What is the reason for this change ?

**Review by mikahanninen — CHANGES_REQUESTED** [review]

File: devdata/env.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (kariharju) [reply devdata/env.json:1]: Yeah env.json is also needed for the rcc users, so I'd keep it.

File: consumer.robot (hunk 5 above)
Comment (kariharju) [reply consumer.robot:13]: As this is a template I would also keep the "catch all" -style.. less questions in support when people just take this and do not know how except really works.

File: producer.robot (hunk 7 above)
Comment (kariharju) [reply producer.robot:14]: These need to be `ROBOT_ARTIFACTS` the stuff that is pushed to Control Room is separated by this and needs to be clear.. robot outputs can go in another directory. 
E.g. customer does not want rfw log.html to Cloud but needs a few files to go there.

File: devdata/env.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (cmin764) [reply devdata/env.json:1]: Isn't this _env.json_ a duplicate of _env-for-producer.json_ one? And VSCode automatically detects now the Work Items, it doesn't require any _env.json_ at all. ([details](https://github.com/robocorp/template-producer-consumer/pull/2#discussion_r899199875))

File: producer.robot (hunk 7 above)
Comment (cmin764) [reply producer.robot:14]: Good to know, there was a strange change before this diff from `%{ROBOT_ARTIFACTS}` -> `%{ROBOT_ARTIFACT}` and from there this change stemmed.

Comment (blakewaud) [reply producer.robot:14]: I have never used nor seen this variable so that was why I changed it because I expected my test to fail because this environment variable isn't defined and I didn't see it in any of the documentation. Are there other such run time variables that we need to define in our documentation that is not included in the Robot Framework user guide?

File: devdata/env.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (blakewaud) [reply devdata/env.json:1]: Okay dokey

File: consumer.robot (hunk 5 above)
Comment (blakewaud) [reply consumer.robot:13]: I disagree, especially with some of the recent discussions about WDM. My understanding is we are going to be changing how an uncaught error releases a work item. A bare except will catch errors that devs didn't account for and cause the robot to technically not fail, but throw an error. In this situation, the final log will be more difficult to troubleshoot. Also, there are risks with a bare except, for example, it will catch `KeywordNotFoundError` and other syntax errors.

**Comment by blakewaud** [comment]
Discussed with @mikahanninen on a call to clarify and simply the template. I will add all the more complex teardown and setup stuff into the web store producer consumer portal bot.
